HW2/P2/parallel_vector.pyx

```diff
@@ -70,23 +70,46 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    cdef:
+        int idx, r, src_index, dest_index, locks_count
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4):
+                src_index = src[idx]
+                dest_index = dest[idx]
+                locks_count = 0
+                
+                #Acquire locks in consistent order prior to checking value > 0
+                if src_index < dest_index:
+                    acquire(&locks[src_index])
+                    acquire(&locks[dest_index])
+                    locks_count = 2
+                elif dest_index < src_index:
+                    acquire(&locks[dest_index])
+                    acquire(&locks[src_index])
+                    locks_count = 2
+                else:
+                    acquire(&locks[src_index])
+                    locks_count = 1
+                
+                #Move value
+                if counts[src_index] > 0:
+                    counts[dest_index] += 1
+                    counts[src_index] -= 1
+                
+                #Release locks
```

Comment:
I had released locks in order as acquiring them but I realized that it is unnecessary since deadlocks cannot happen in the case of releasing two locks in any order.